contracts/src/v0.8/keystone/KeystoneForwarder.sol

```diff
@@ -2,75 +2,227 @@
+    s_configs[donId].f = f;
+  }
+
+  // send a report to receiver
   function report(
-    address targetAddress,
-    bytes calldata data,
+    address receiverAddress,
+    bytes calldata rawReport,
     bytes[] calldata signatures
-  ) external nonReentrant returns (bool) {
-    if (data.length < SELECTOR_LENGTH + REPORT_LENGTH) {
-      revert InvalidData(data);
+  ) external nonReentrant {
+    if (rawReport.length < REPORT_METADATA_LENGTH) {
+      revert InvalidReport();
     }
 
-    // data is an encoded call with the selector prefixed: (bytes4 selector, bytes report, ...)
-    // we are able to partially decode just the first param, since we don't know the rest
-    bytes memory rawReport = abi.decode(data[4:], (bytes));
+    (bytes32 workflowId, uint32 donId, bytes32 workflowExecutionId, address workflowOwner) = _getMetadata(rawReport);
+
+    // f can never be 0, so this means the config doesn't actually exist
+    if (s_configs[donId].f == 0) revert InvalidDonId(donId);
 
-    // TODO: we probably need some type of f value config?
+    bytes32 reportId = _reportId(receiverAddress, workflowExecutionId);
+    if (s_reports[reportId].transmitter != address(0)) revert ReportAlreadyProcessed(reportId);
 
-    bytes32 hash = keccak256(rawReport);
+    if (s_configs[donId].f + 1 != signatures.length)
+      revert WrongNumberOfSignatures(s_configs[donId].f + 1, signatures.length);
 
     // validate signatures
-    for (uint256 i = 0; i < signatures.length; i++) {
-      // TODO: is libocr-style multiple bytes32 arrays more optimal?
-      (bytes32 r, bytes32 s, uint8 v) = Utils._splitSignature(signatures[i]);
-      address signer = ecrecover(hash, v, r, s);
-      // TODO: we need to store oracle cluster similar to aggregator then, to validate valid signer list
+    {
+      bytes32 hash = keccak256(rawReport);
+
+      address[MAX_ORACLES] memory signed;
+      uint8 index = 0;
+      for (uint256 i = 0; i < signatures.length; i++) {
+        // TODO: is libocr-style multiple bytes32 arrays more optimal, gas-wise?
+        (bytes32 r, bytes32 s, uint8 v) = _splitSignature(signatures[i]);
+        address signer = ecrecover(hash, v, r, s);
+
+        // validate signer is trusted and signature is unique
+        index = uint8(s_configs[donId]._positions[signer]);
+        if (index == 0) revert InvalidSigner(signer); // index is 1-indexed so we can detect unset signers
+        index -= 1;
+        if (signed[index] != address(0)) revert DuplicateSigner(signer);
+        signed[index] = signer;
+      }
     }
 
-    (bytes32 workflowId, bytes32 workflowExecutionId) = Utils._splitReport(rawReport);
-
-    // report was already processed
-    if (s_reports[workflowExecutionId] != address(0)) {
-      return false;
+    IReceiver receiver = IReceiver(receiverAddress);
+    bool success;
+    try receiver.onReport(workflowId, workflowOwner, rawReport[REPORT_METADATA_LENGTH:]) {
+      success = true;
+    } catch {
+      success = false;
```

Comment:
Should we emit an event for alerting here?

Reply:
The event is emitted below.